### DISCUSSION TITLE
Support restructured text in Pyright language server

### DISCUSSION THREAD

**Discussion by Alan-Manning** [microsoft/pylance-release#6631] [Enhancement]
**Explanation and Screenshot**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

Pyright does not provide the correct formatting for docstrings of the numpy type format, see pic below.

![Image](https://github.com/user-attachments/assets/9cc0226b-be22-4929-bae7-67577311d22d)

The syntax has escapes for all the underscors, square brackets. Also there is an issue with the indentation in the hoverdocs where the indentation is correct in the docstring but it gets ignored past the first line for each parameter, kwarg etc.

This is with pyright installed in Neovim. using Mason to install and manage it. Using nvim-lspconfig to configure the language sever. The hover docs provider is the one provided by Nvim.

I have looked through the lsplogs and found that the text sent to the hover docs is exactly as written out here including the slashes infront of the underscores etc. The only thing that Nvim is doing is replacing the &nbsp; characters with whitespace untill visually selected. pic below shows the visual selection over that range showing the &nbsp chars.

 ![Image](https://github.com/user-attachments/assets/9ece2317-9415-4694-b6ce-96fec569774d)

Ideally the text sent to the hover docs would only have the whitespace chars instead of the &nbsp; , or at least it should maintain the use of the &nbsp; to continue the indentation inline with what is in the docstring. Also the misc slashes should be gone.

Please let me know if i should provide any more information to help solv but i believe this is a bug that should be fixed within pyright rather than a Nvim thing.

 Thank you
 - Alan Manning

**Comment by erictraut** [comment]
The Microsoft Pylance team owns and maintains the docstring formatting functionality that is included in pyright.

Could someone from the pylance team please transfer this issue to the [pylance-release project](https://github.com/microsoft/pylance-release)? Thanks!

**Comment by rchiodo** [comment]
Restructured text is only currently supported in Pylance (inside of VS code). 

Moving this to a discussion item for upvotes.

**Comment by jdhao** [comment]
I see the same issue. In the original code documentation, the plain white spaces are converted to `&nbsp;` by pyright. Not sure why?

**Reply by rchiodo** [reply]
They are converted so they are formatted as actual spaces by the markdown renderer in VS code. Perhaps Pyright isn't handling the format flag correctly here or Neovim is passing the 'markdown' flag for hover text formatting when it should be using plain text.

**Reply by Alan-Manning** [reply]
i mean that is fine if it is sending something that is correctly rendered in the end. maybe it sends &nobsp; and i can just configure nvim to not show that even on hover, this is reasonable. however there is still the issue of the escaped chars with the slash and most annoying is that whatever mehtod is used for indent, it doesnt respect the indentation in the file.

**Reply by rchiodo** [reply]
Yeah, that would be plain text mode. Plain text mode is supposed to return the same text as in the docstring as the hover text. Indents would be the same and all spaces would be just spaces.

**Reply by Alan-Manning** [reply]
where could one find docs on the plain text mode?

**Reply by jdhao** [reply]
> where could one find docs on the plain text mode?

Hi, I have replied to your original reddit post in neovim reddit. Please check there 🙏

**Reply by Alan-Manning** [reply]
https://www.reddit.com/r/neovim/comments/1gdv1rc/what_is_causeing_the_lsp_hover_docs_to_looks_like/

just so this thread has this link i guess

**Comment by Alan-Manning** [comment]
i still feel that there should be some other option for formatting that includes some regular markdown view.

**Reply by rchiodo** [reply]
I believe `&nbsp;` is a pretty [standard ](https://stackoverflow.com/questions/15721373/how-do-i-ensure-that-whitespace-is-preserved-in-markdown) way to get markdown to render spaces. You might ask NeoVIM to support it in hover.

**Reply by Alan-Manning** [reply]
so what is the main reason for the docstring not matching what is written in the code. i.e what is causing the indentation to be incorrect when you use the markdown opposed to the plain text?

**Reply by rchiodo** [reply]
If the IDE asks for markdown, and we want to preserve spacing (as HTML removes spaces), we need some way to represent a space. The most common way to do so is with a `&nbsp;` embedded in the markdown. 

For example, if you output this as html:

```text
        Indented text
```

It will render like so:

```text
Indented text
```

So if you want to preserve the indent, you need to add something special to the output. One of the ways to do that is to render spaces with `&nbsp;` Another way is to use `<ul>` tags (we do that in the Pylance side to indent parts of the docstring).

However, if the IDE is not rendering markdown, but just rendering text, these extra bits don't work.

**Reply by Alan-Manning** [reply]
yeah that is fine that there neeeds to be something to represent blank space, The nbsp is being rendered as a space untill i hover over it, that is something i can change within some custom lua func to replace that. but its the next lines, so i think its the extra bits like the `<ul>` that dont get rendered. And i am assuming that the ul list tag is used for subsequent indented lines like in pic below?
![image](https://github.com/user-attachments/assets/f46609d4-9c37-4a25-b00b-04cfd0597bea)

if this isnt the case then what is the characters that should be getting output infront of those lines in your markdown such that the indentation is preserved? I have tried printing the raw output and i do not see any of these special chars. 
![image](https://github.com/user-attachments/assets/892a1d35-0060-4675-95ba-6ed706076bb6)

i think its the case that im not being given these other special chars in the first place. i can post some of the docstring here to see if you get this formatted correctly, and perhaps if so you could provide the raw text that is sent to the hover window to get rendered?

```python
def add_test_H_pad_connected_box_section(
    x: float | int,
    y: float | int,
    add_dielectric_cutout_over_line: bool | Sequence[bool] = (False, False, False),
) -> None:
    """Adds a box of size 5500 centered on the x,y given that contains a series
    of 6 pads where 2 of those connect to a line connecting them.

    Parameters
    ----------
    mask: SoukMask
        SoukMask.

    x, y: float, int
        The x, y coordinate to center the test structure.

    pad_materials: Layer | Sequence[Layer]
        This is the layers that each of the pad pairs should be drawn in.
        This is an instance of Layer of a Sequence of many Layers. If a single
        Layer is given then all the ports will be drawn in this layer.
        If a Sequence of Layers is given then it should be of length 6 and each
        pad pair will take the respective layer from left to right. For more info
        see maskpy.layers.Layer. Usually this is within the
        SoukMaskBuilder.layers.xxx. e.g. `mask.layers.Aluminium`

    line_materials: Layer | Sequence[Layer]
        This is the layers that each of the lines should be drawn in.
        This is an instance of Layer of a Sequence of many Layers. If a single
        Layer is given then all the lines will be drawn in this layer.
        If a Sequence of Layers is given then it should be of length 3 and each
        of the lines will take the respective layer from left to right. For more
        info see maskpy.layers.Layer. Usually this is within the
        SoukMaskBuilder.layers.xxx. e.g. `mask.layers.Aluminium`

    KwArgs
    ------
    rot: float | int = 0
        The rotation for the entire structure given in radians.

    linewidths: float | int = 2.0,
        This is the linewidths to use for each of the lines. Default is 2.0
        which will draw the lines all of this linewidth. This wil be the same
        for the U shape connections leading up to the connecting line.
    """
    return None
```

**Reply by rchiodo** [reply]
For that docstring, here's what we return:

```
2024-11-15 09:21:17.647 [info] (26984) [FG] binding: file:///c%3A/Users/rchiodo/source/testing/test_pylance/test_docstring.py (0ms)
2024-11-15 09:21:17.649 [info] [Trace - 9:21:17 AM] Received response 'textDocument/hover - (84)' in 372ms.
2024-11-15 09:21:17.649 [info] Result: {
    "contents": {
        "kind": "markdown",
        "value": "```python\n(function) def add_test_H_pad_connected_box_section(\n    x: float | int,\n    y: float | int,\n    add_dielectric_cutout_over_line: bool | Unknown = (False, False, False)\n) -> None\n```\n---\nAdds a box of size 5500 centered on the x,y given that contains a series\nof 6 pads where 2 of those connect to a line connecting them.\n\nParameters\n----------\nmask: SoukMask  \n&nbsp;&nbsp;&nbsp;&nbsp;SoukMask.\n\nx, y: float, int  \n&nbsp;&nbsp;&nbsp;&nbsp;The x, y coordinate to center the test structure.\n\npad\\_materials: Layer | Sequence\\[Layer\\]  \n&nbsp;&nbsp;&nbsp;&nbsp;This is the layers that each of the pad pairs should be drawn in.\nThis is an instance of Layer of a Sequence of many Layers. If a single\nLayer is given then all the ports will be drawn in this layer.\nIf a Sequence of Layers is given then it should be of length 6 and each\npad pair will take the respective layer from left to right. For more info\nsee maskpy.layers.Layer. Usually this is within the\nSoukMaskBuilder.layers.xxx. e.g. `mask.layers.Aluminium`\n\nline\\_materials: Layer | Sequence\\[Layer\\]  \n&nbsp;&nbsp;&nbsp;&nbsp;This is the layers that each of the lines should be drawn in.\nThis is an instance of Layer of a Sequence of many Layers. If a single\nLayer is given then all the lines will be drawn in this layer.\nIf a Sequence of Layers is given then it should be of length 3 and each\nof the lines will take the respective layer from left to right. For more\ninfo see maskpy.layers.Layer. Usually this is within the\nSoukMaskBuilder.layers.xxx. e.g. `mask.layers.Aluminium`\n\nKwArgs\n------\nrot: float | int = 0  \n&nbsp;&nbsp;&nbsp;&nbsp;The rotation for the entire structure given in radians.\n\nlinewidths: float | int = 2.0,  \n&nbsp;&nbsp;&nbsp;&nbsp;This is the linewidths to use for each of the lines. Default is 2.0\nwhich will draw the lines all of this linewidth. This wil be the same\nfor the U shape connections leading up to the connecting line.\n<!--moduleHash:-1984330094-->"
    },
    "range": {
        "start": {
            "line": 2072,
            "character": 4
        },
        "end": {
            "line": 2072,
            "character": 40
        }
    }
}
```

In VS code that renders like this (when using Pyright):

![image](https://github.com/user-attachments/assets/2bab8951-bfd7-4a43-8cb3-79d4966a1af0)

**Reply by rchiodo** [reply]
I believe Pyright is not going to give back well formatted markdown for that text no matter what transforms you do. It doesn't support restructured text well.

**Reply by Alan-Manning** [reply]
is that something that could be implimented in the future or should i switch away from pyright in order to get this working correctly?

**Reply by rchiodo** [reply]
Pylance supports it at the moment (meaning you have to use VS code). This discussion item is to see how many people are asking for this support outside of VS code. 

My guess is no, this won't be implemented in Pyright.

**Reply by Alan-Manning** [reply]
so i though that pyright was using pylance in order to provide the hover docs. what about the responce from pylance differs if you use vscode over any other editor? if the responce is the same then what is vscode doing to that to make it format nicely?

**Reply by rchiodo** [reply]
Sorry but pyright and pylance are not using the same logic. We have custom logic in Pylance only that provides way more parsing around reST output. And that parsing then does a bunch of extra work to generate appropriate markdown output. 

It was originally done this way because we didn't know if it would work and we didn't want to break Pyright's existing hover documentation. 

The question now is whether or not it's worth pushing this new hover provider down to Pyright. It's better in some regards but still has bugs in it:
https://github.com/microsoft/pylance-release/issues?q=is%3Aissue%20state%3Aopen%20label%3Adocstrings

**Reply by Alan-Manning** [reply]
I mean i think that this should be pushed down under some experimental flag/ something that is opt in to enable. There are bugs in most things so i dont think excluding it from pyright becuase it has bugs is really that bad and from what i can see of those issues they dont really interact with the markdown view in the normal cases. I dont know the complexity of implimenting the pylance logic into the pyright hover docs but would be nice to have the option if the code already exists for the more complex restructred text parsing.

**Comment by Alan-Manning** [comment]
has there been any update for this topic?

**Reply by rchiodo** [reply]
Sorry but no updates. It's not on our plan at the moment.
